les/api.go

```diff
@@ -17,16 +17,408 @@
+			delete(api.subs, sub)
+		case <-sub.notifier.Closed():
+			delete(api.subs, sub)
+		default:
+			sub.notifier.Notify(sub.rpcSub.ID, event)
+		}
+	}
+}
+
+// SetClientParams sets client parameters for all clients listed in the ids list or matching the given tags
+func (api *PrivateLightServerAPI) SetClientParams(ids []enode.ID, params map[string]interface{}) error {
+	updateFactorsWithParam := func(name string, value interface{}, posFactors, negFactors *priceFactors) (bool, error) {
+		var err error
+		switch name {
+		case "pricing/timeFactor":
+			if val, ok := value.(float64); ok && val >= 0 {
+				posFactors.timeFactor = val / 1000000000
+			} else {
+				err = errInvalidValue
+			}
+		case "pricing/capacityFactor":
+			if val, ok := value.(float64); ok && val >= 0 {
+				posFactors.capacityFactor = val / 1000000000
+			} else {
+				err = errInvalidValue
+			}
+		case "pricing/requestCostFactor":
+			if val, ok := value.(float64); ok && val >= 0 {
+				posFactors.requestFactor = val / 1000000000
+			} else {
+				err = errInvalidValue
+			}
+		case "pricing/negative/timeFactor":
+			if val, ok := value.(float64); ok && val >= 0 {
+				negFactors.timeFactor = val / 1000000000
+			} else {
+				err = errInvalidValue
+			}
+		case "pricing/negative/capacityFactor":
+			if val, ok := value.(float64); ok && val >= 0 {
+				negFactors.capacityFactor = val / 1000000000
+			} else {
+				err = errInvalidValue
+			}
+		case "pricing/negative/requestCostFactor":
+			if val, ok := value.(float64); ok && val >= 0 {
+				negFactors.requestFactor = val / 1000000000
+			} else {
+				err = errInvalidValue
+			}
+		default:
+			return false, nil
+		}
+		return err == nil, err
+	}
+
+	var finalErr error
+	if len(ids) == 0 {
+		// update default price factors
+		var update bool
+		for name, value := range params {
+			u, err := updateFactorsWithParam(name, value, &api.defaultPosFactors, &api.defaultNegFactors)
+			update = update || u
+			if err != nil {
+				if finalErr == nil || finalErr == err {
+					finalErr = err
+				} else {
+					finalErr = errMultiple
+				}
+			}
+		}
+		if update {
+			api.server.clientPool.setDefaultFactors(api.defaultPosFactors, api.defaultNegFactors)
+		}
+	}
+	api.server.clientPool.forClients(ids, func(client *clientInfo, id enode.ID) {
+		var (
+			err            error
+			updatePrice, u bool
+		)
+		for name, value := range params {
+			if client != nil {
+				if u, err = updateFactorsWithParam(name, value, &client.posFactors, &client.negFactors); u {
+					updatePrice = true
+				}
+			} else {
+				var dummy priceFactors
+				if u, err = updateFactorsWithParam(name, value, &dummy, &dummy); u {
+					err = errClientNotConnected
+				}
+			}
+			if !u && err == nil {
+				switch name {
+				case "capacity":
+					if client != nil {
+						if capacity, ok := value.(float64); ok && (capacity == 0 || uint64(capacity) >= api.server.minCapacity) {
+							err = api.server.clientPool.setCapacity(client, uint64(capacity))
```

Comment:
this is an `if` in an `if` in a  `case` in a `switch` in a `for` in an anonymous `func`.  Can this be rewritten for clarity?

Reply:
yes, done